cmd/atlas/internal/cmdext/cmdext.go

```diff
@@ -192,3 +201,134 @@ func blockError(name string, b *hclsyntax.Block) func(string, ...any) error {
 		return fmt.Errorf("%s.%s: %w", name, b.Labels[1], fmt.Errorf(format, args...))
 	}
 }
+
+type (
+	// LoadStateOptions for external state loaders.
+	LoadStateOptions struct {
+		URLs []*url.URL
+		Dev  *sqlclient.Client // Client for the dev database.
+	}
+	// StateLoader allows loading StateReader's from external sources.
+	StateLoader interface {
+		LoadState(context.Context, *LoadStateOptions) (migrate.StateReader, error)
+	}
+
+	// MigrateDiffOptions for external migration differ.
+	MigrateDiffOptions struct {
+		Name string
+		To   []string
+		Dir  migrate.Dir
+		Dev  *sqlclient.Client
+	}
+	// MigrateDiffer allows external sources to implement custom migration differs.
+	MigrateDiffer interface {
+		MigrateDiff(context.Context, *MigrateDiffOptions) error
+		needDiff([]string) bool
+	}
+)
+
+// States is a global registry for external state loaders.
+var States = registry{
+	"ent": EntLoader{},
+}
+
+type registry map[string]StateLoader
+
+// Loader returns the state loader for the given scheme.
+func (r registry) Loader(scheme string) (StateLoader, bool) {
+	l, ok := r[scheme]
+	return l, ok
+}
+
+// Differ returns the raw states differ for the given URLs, if registered.
+func (r registry) Differ(to []string) (MigrateDiffer, bool) {
+	for _, l := range r {
+		if d, ok := l.(MigrateDiffer); ok && d.needDiff(to) {
+			return d, true
+		}
+	}
+	return nil, false
+}
+
+// EntLoader is a StateLoader for loading ent.Schema's as StateReader's.
+type EntLoader struct{}
+
+// LoadState returns a migrate.StateReader that reads the schema from an ent.Schema.
+func (l EntLoader) LoadState(ctx context.Context, opts *LoadStateOptions) (migrate.StateReader, error) {
+	switch {
+	case opts.Dev == nil, len(opts.URLs) != 1:
+		return nil, errors.New("schema url and dev database are required")
+	case opts.URLs[0].Query().Has("globalid"):
+		return nil, errors.New("globalid is not supported by this command. Use 'migrate diff' instead")
```

Comment:
This confuses me. Not sure if miss some context but this shouldn't be here, right?

Reply:
Ent schema with `globalid` is supported only by `migrate diff` and should be rejected otherwise, because Atlas doesn’t support appending values into the `ent_types` table.

Reply:
Ah, now I got it.